### PR TITLE
fix: handle nil Executor.Instances in GetExecutorRequestResource

### LINKED ISSUE #2832: `GetExecutorRequestResource`  panics when dynamic allocation is enabled with batch scheduler
### What happened?

- [x] ✋ I have searched the open/closed issues and my issue is not listed.

Controller panics at https://github.com/kubeflow/spark-operator/blob/2e96ece15939cd7eb365b34268f416ae90f67ec9/pkg/util/sparkapplication.go#L469
 when a SparkApplication uses dynamic allocation with a batch scheduler (example Volcano), line 469 dereferences `*app.Spec.Executor.Instances` without nil check. When dynamic allocation is enabled, `Instances` remains nil because the defaulting logic at 
https://github.com/kubeflow/spark-operator/blob/2e96ece15939cd7eb365b34268f416ae90f67ec9/api/v1beta2/defaults.go#L95-L98
 only sets it when dynamic allocation is disabled.

### Reproduction Code

1. Install spark-operator with `--set controller.batchScheduler.enable=true`
2. Install Volcano scheduler
3. Apply this SparkApplication
```yaml
apiVersion: sparkoperator.k8s.io/v1beta2
kind: SparkApplication
metadata:
  name: test-dynamic-alloc
spec:
  type: Scala
  mode: cluster
  image: spark:3.5.0
  mainClass: org.apache.spark.examples.SparkPi
  mainApplicationFile: local:///opt/spark/examples/jars/spark-examples.jar
  sparkVersion: "3.5.0"
  batchScheduler: volcano
  dynamicAllocation:
    enabled: true
    minExecutors: 1
    maxExecutors: 10
  driver:
    cores: 1
    memory: "512m"
    serviceAccount: spark
  executor:
    cores: 1
    memory: "512m"
```

### Expected behavior

controller handles nil `Executor.Instances` gracefully 

### Actual behavior

```bash
panic: runtime error: invalid memory address or nil pointer dereference
github.com/kubeflow/spark-operator/v2/pkg/util.GetExecutorRequestResource
    /workspace/pkg/util/sparkapplication.go:469
```

### Environment & Versions

- Kubernetes Version: v1.30.0
- Spark Operator Version: v2.3.0
- Apache Spark Version: 3.5.0

### Additional context

_No response_

### Impacted by this bug?

Give it a 👍 We prioritize the issues with most 👍

### PR DESCRIPTION
**Pull request by puwun**
Fixes #2832

## Purpose of this PR

When dynamic allocation is enabled `Executor.Instances` is intentionally left as `nil` by the defaulting logic in 
`api/v1beta2/defaults.go`. However, `GetExecutorRequestResource` dereferences this field without a nil check, causing a panic when used with batch schedulers (like Volcano/kube-scheduler)

**Proposed changes:**
- Add nil check for `Executor.Instances` in `GetExecutorRequestResource` function
- Return empty `ResourceList{}` when Instances is `nil` 
- Add unit tests covering both nil and non-nil Instances scenarios

## Change Category

- [x] Bugfix (non-breaking change which fixes an issue)

## Checklist

- [x] I have conducted a self-review of my own code.
- [x] I have updated documentation accordingly.
- [x] I have added tests that prove my changes are effective or that my feature works.
- [x] Existing unit tests pass locally with my changes.

### Additional Notes
Verified in Kind cluster with Volcano scheduler - SparkApplication with `dynamicAllocation.enabled: true` and `batchScheduler: volcano` now processes successfully without panic

### CODE REVIEW

File: pkg/util/sparkapplication.go [hunk 1]
```diff
@@ -465,6 +465,11 @@ func GetExecutorRequestResource(app *v1beta2.SparkApplication) corev1.ResourceLi
 		}
 	}
 
+	// Handle nil Instances (e.g., when dynamic allocation is enabled)
+	if app.Spec.Executor.Instances == nil {
+		return corev1.ResourceList{}
+	}
+
```
Comment (ChenYi015) [inline pkg/util/sparkapplication.go]: When dynamic allocation is enabled, it is better to return the initial resource list rather than an empty list. The initial number of executors is decided by `spark.dynamicAllocation.initialExecutors`, `spark.executor.instances` (i.e. `app.Spec.Executor.Instances`), see https://spark.apache.org/docs/latest/configuration.html#dynamic-allocation.
